virtual_rainforest/core/model.py

```diff
@@ -1,12 +1,42 @@
-"""The `core.model` module.
-
-The `core.model` module defines the api that all individual models (e.g. the soil
-model) should conform to. This consists of a class (`Model`), which defines the expected
-functions. The relevant modules will create classes to represent specific models, which
-will inherit from the `Model` base class. These subclasses will generally overwrite the
-functions defined in the base class, which are defined mainly to force a consistent api
-between models. It also establishes a model registry that allows models to become
-accessible across scripts without individual loading in.
+"""API documentation for the :mod:`core.model` module.
+************************************************** # noqa: D205
+
+The `core.model` module defines the api that all individual models (e.g. the soil model)
+should conform to. This consists of a class (`BaseModel`), which defines the expected
+functions. This class is an abstract base class, which is a class that is never intended
+to be instantiated itself but instead serves as a blueprint for inheriting classes. Some
+of its functions will be inherited and used by its child classes (e.g. `__repr__` and
+`__str__`), unless they are explicitly overwritten in the child class. However, it also
+possesses abstract methods (denoted by `@abstractmethod`), which are merely
+placeholders. For these abstract methods child classes have to overwrite the methods
+with new functions, otherwise class inheritance fails. This ensures that a consistent
+api (set of functions) is used across models, while also ensuring that functions don't
+default to an inappropriate generic behaviour due to a function not being defined for a
+particular model. Abstract classes should thus be used for functions that are always
```

Comment:
do you mean 'abstract class' or 'abstract method' here?
Also, does that mean that every function we write in our modules needs to be a child of the abstract method class?

Reply:
I think the idea is that the high-level functions that setup, pinup and run the whole model should all follow a consistent pattern. Think these functions will call loads of module specific functions. I don't think there's any sense in the module specific functions being based on an abstract base class in general, as there won't be great similarities between modules at this level. Basically what `setup`, `spinup`, `solve` etc are there to do is to transition from the shared high level api to the module specific details, so as far as I'm aware I don't think there's any sense in lower level functions sharing a common framework/abc.

Reply:
And yeah you are right that should be `Abstract methods`

Reply:
Do you think I should add any of my above reply into the docstring?

Reply:
Yes, I think it would be good if you could clarify this paragraph a bit more 👍

Reply:
I think the more examples you can provide, the better. It is quite an abstract topic...